src/services/login-util.js

```diff
@@ -0,0 +1,139 @@
+const http = require('http')
+const crypto = require('crypto')
+const consola = require('consola')
+const axios = require('axios')
+const config = require('./config')
+
+const AUTH0_DOMAIN = 'checkly'
+const AUTH0_CLIENT_ID = 'mBtwLFVm39GVZ1HpSRBSdRiLFucYxmMb'
+
+const generateAuthenticationUrl = (codeChallenge, scope, state) => {
+  const url = new URL(`https://${AUTH0_DOMAIN}.eu.auth0.com/authorize`)
+  const params = new URLSearchParams({
+    client_id: AUTH0_CLIENT_ID,
+    code_challenge: codeChallenge,
+    code_challenge_method: 'S256',
+    response_type: 'code',
+    redirect_uri: 'http://localhost:4242',
+    scope: scope,
+    state: state,
+  })
+
+  url.search = params
+  return url.toString()
+}
+
+function generatePKCE() {
+  const codeVerifier = crypto
+    .randomBytes(64)
+    .toString('base64')
+    .replace(/\+/g, '-')
+    .replace(/\//g, '_')
+    .replace(/=+$/, '')
+
+  const codeChallenge = crypto
+    .createHash('sha256')
+    .update(codeVerifier)
+    .digest('base64')
+    .replace(/\+/g, '-')
+    .replace(/\//g, '_')
+    .replace(/=+$/, '')
+
+  return {
+    codeChallenge,
+    codeVerifier,
+  }
+}
+
+function startServer(getToken) {
+  const server = http.createServer()
+  server.on('request', (req, res) => {
+    // `req.url` has a '/' char at the beginning which needs removed to be valid searchParams input
+    const responseParams = new URLSearchParams(req.url.substring(1))
+    const code = responseParams.get('code')
+    const state = responseParams.get('state')
+
+    if (code && state) {
+      res.write(`
+      <html>
+      <body>
+        <div style="height:100%;width:100%;inset:0;position:absolute;display:grid;place-items:center;background-color:#EFF2F7;text-align:center;font-family:Inter;">
```

Comment:
As we discussed, I think the Vercel cli directly exits when the auth is successful

Reply:
Ah yes, implemented! :+1:

Reply:
nice